src/components/NewBranchDialog.tsx

```diff
@@ -0,0 +1,440 @@
+import * as React from 'react';
+import { classes } from 'typestyle';
+import List from '@material-ui/core/List';
+import ListItem from '@material-ui/core/ListItem';
+import Dialog from '@material-ui/core/Dialog';
+import DialogActions from '@material-ui/core/DialogActions';
```

Comment:
Using dialog component from third-party library breaks the homogeneity of the code. But in the other hand this eases code development. @telamonian what is your opinion on this?

Note: personally, if the dialog look is similar to the default JupyterLab dialog, I'm fine with it.

Reply:
In general, I recommend moving away from JupyterLab dialogs. If anything, the use of JupyterLab [dialogs](https://github.com/jupyterlab/jupyterlab-git/blob/576c4448f2ce19c63604a949e343da498cf81858/src/widgets/gitPushPull.ts#L14) breaks homogeneity of the extension code (i.e., React components), not the other way around.

Upon initially submitting the PR, I ran the use of Material UI dialogs past @telamonian and he thought this to be fine.

Reply:
It's not that I wouldn't prefer a home grown, tastefully themed solution, but there's some problems:

- The Dialog box from @jupyterlab/apputils just isn't that flexible. It hard to significantly customize or use in complex scenarios.

- The apputils Dialog also isn't React based. The way it implements events (like what happens when you click a button) aren't naturally compatible with standard React patterns.

Because of the above, I once wasted an entire evening trying to implement a custom form using the Dialog from jlab core. I wanted to open the dialog box using one React component that was nested within a tree, then update the top-level component in the tree based on the user's input. It was not working out. On the other hand, once I decided to switch to the Dialog from material-ui, I had it solved within about 15 min.

I definitely see your point Frederic about homogeneity, and about the value of keeping things simple and consistent. It is true that material-ui brings along its own complex config system (with a rather foreign set of conventions). However, I don't think it makes sense for us to favor any particular approach based on a primary argument of homogeneity, or "because it's provided by jlab core". Instead, we should ask ourselves which component will allow us to build the most kick-ass and easily maintained `jupyterlab-git` that we can manage. The Dialog component from material-ui is simply much better than the Dialog from jlab core, even more so since we're primarily working within React here.

So based on my own experience (and I suppose I am still a bit salty about the apputils Dialog wasting my time), I think that Athan absolutely made the right call here to use material-ui. My only caveat is the same as Frederic's: the theming overall needs to mesh with that of the rest of jlab (though it certainly doesn't have to perfectly match).

On the horizon of a month or two from now, a React-based dialog box is definitely on my list of things to add to @jupyterlab/ui-components. Once that exists, maybe it will be appropriate to revisit this discussion.

Reply:
To your point @telamonian, I tried to style the dialog such that it does mesh, using theme classes, etc. So we should be good on that front.